File: backend/src/main/java/com/gxs/bank/service/AuthService.java
package com.gxs.bank.service;

import com.gxs.bank.dto.request.LoginRequest;
import com.gxs.bank.dto.request.RegisterRequest;
import com.gxs.bank.dto.response.AuthResponse;
import com.gxs.bank.exception.BadRequestException;
import com.gxs.bank.exception.ResourceNotFoundException;
import com.gxs.bank.model.User;
import com.gxs.bank.repository.UserRepository;
import com.gxs.bank.security.JwtTokenProvider;
import lombok.RequiredArgsConstructor;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.UUID;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtTokenProvider tokenProvider;

    public AuthResponse register(RegisterRequest request) {
        if (userRepository.existsByEmail(request.getEmail())) {
            throw new BadRequestException("Email already registered");
        }

        User.Role role;
        try {
            role = User.Role.valueOf(request.getRole().toUpperCase());
        } catch (IllegalArgumentException e) {
            throw new BadRequestException("Invalid role. Must be CUSTOMER or EMPLOYEE");
        }

        User.EmployeeRole employeeRole = null;
        if (role == User.Role.EMPLOYEE && request.getEmployeeRole() != null) {
            try {
                employeeRole = User.EmployeeRole.valueOf(request.getEmployeeRole().toUpperCase());
            } catch (IllegalArgumentException e) {
                throw new BadRequestException("Invalid employee role. Must be MAKER, CHECKER, or ADMIN");
            }
        }

        User user = User.builder()
                .fullName(request.getFullName())
                .email(request.getEmail())
                .phone(request.getPhone())
                .passwordHash(passwordEncoder.encode(request.getPassword()))
                .role(role)
                .employeeRole(employeeRole)
                .employeeId(request.getEmployeeId())
                .department(request.getDepartment())
                .build();

        user = userRepository.save(user);
        return buildAuthResponse(user);
    }

    public AuthResponse login(LoginRequest request) {
        User user = userRepository.findByEmail(request.getEmail())
                .orElseThrow(() -> new BadRequestException("Invalid email or password"));

        if (!passwordEncoder.matches(request.getPassword(), user.getPasswordHash())) {
            throw new BadRequestException("Invalid email or password");
        }

        return buildAuthResponse(user);
    }

    public User getUserById(UUID userId) {
        return userRepository.findById(userId)
                .orElseThrow(() -> new ResourceNotFoundException("User not found"));
    }

    private AuthResponse buildAuthResponse(User user) {
        String token = tokenProvider.generateToken(user.getId(), user.getEmail(), user.getRole().name());

        return AuthResponse.builder()
                .token(token)
                .type("Bearer")
                .userId(user.getId())
                .fullName(user.getFullName())
                .email(user.getEmail())
                .phone(user.getPhone())
                .role(user.getRole().name())
                .kycStatus(user.getKycStatus() != null ? user.getKycStatus().name() : "UNVERIFIED")
                .employeeRole(user.getEmployeeRole() != null ? user.getEmployeeRole().name() : null)
                .department(user.getDepartment())
                .employeeId(user.getEmployeeId())
                .onboardingStatus(user.getOnboardingStatus() != null ? user.getOnboardingStatus().name() : null)
                .build();
    }
}

File: backend-python/com/gxs/bank/service/AuthService.py
from dataclasses import asdict

from com.gxs.bank.config.SecurityConfig import passwordEncoder, passwordMatches
from com.gxs.bank.dto.request.LoginRequest import LoginRequest
from com.gxs.bank.dto.request.RegisterRequest import RegisterRequest
from com.gxs.bank.dto.response.AuthResponse import AuthResponse
from com.gxs.bank.exception.BadRequestException import BadRequestException
from com.gxs.bank.exception.ResourceNotFoundException import ResourceNotFoundException
from com.gxs.bank.model.User import User
from com.gxs.bank.repository.UserRepository import UserRepository
from com.gxs.bank.security.JwtTokenProvider import JwtTokenProvider


class AuthService:
    def __init__(self, db):
        self.db = db
        self.userRepository = UserRepository(db)
        self.tokenProvider = JwtTokenProvider()

    def register(self, request: RegisterRequest) -> AuthResponse:
        if self.userRepository.existsByEmail(request.email):
            raise BadRequestException("Email already registered")

        try:
            role = User.Role(request.role.upper())
        except ValueError:
            raise BadRequestException("Invalid role. Must be CUSTOMER or EMPLOYEE") from None

        employeeRole = None
        if role == User.Role.EMPLOYEE and request.employeeRole is not None:
            try:
                employeeRole = User.EmployeeRole(request.employeeRole.upper())
            except ValueError:
                raise BadRequestException("Invalid employee role. Must be MAKER, CHECKER, or ADMIN") from None

        user = User(
            fullName=request.fullName,
            email=request.email,
            phone=request.phone,
            passwordHash=passwordEncoder(request.password),
            role=role,
            employeeRole=employeeRole,
            employeeId=request.employeeId,
            department=request.department,
        )
        self.userRepository.save(user)
        self.db.commit()
        self.db.refresh(user)
        return self._buildAuthResponse(user)

    def login(self, request: LoginRequest) -> AuthResponse:
        user = self.userRepository.findByEmail(request.email)
        if user is None or not passwordMatches(request.password, user.passwordHash):
            raise BadRequestException("Invalid email or password")
        return self._buildAuthResponse(user)

    def getUserById(self, userId: str):
        user = self.userRepository.findById(userId)
        if user is None:
            raise ResourceNotFoundException("User not found")
        return user

    def _buildAuthResponse(self, user) -> AuthResponse:
        token = self.tokenProvider.generateToken(user.id, user.email, user.role.value)
        return AuthResponse(
            token=token,
            type="Bearer",
            userId=user.id,
            fullName=user.fullName,
            email=user.email,
            phone=user.phone,
            role=user.role.value,
            kycStatus=user.kycStatus.value if user.kycStatus else "UNVERIFIED",
            employeeRole=user.employeeRole.value if user.employeeRole else None,
            department=user.department,
            employeeId=user.employeeId,
            onboardingStatus=user.onboardingStatus.value if user.onboardingStatus else None,
        )

    def buildAuthResponseDict(self, user):
        return asdict(self._buildAuthResponse(user))
